### PR TITLE
Auto protocol match support xprotocol

### PR DESCRIPTION
**Pull request by cadeeper**
auto match support xprotocol

### UT result
截屏2020-12-03 下午4 21 32

### CODE REVIEW

File: pkg/stream/xprotocol/factory.go [hunk 1]
```diff
@@ -49,5 +50,23 @@ func (f *streamConnFactory) CreateBiDirectStream(context context.Context, connec
...
+		result := matcher(magic)
+		if result == types.MatchSuccess {
+			return nil
+		}
+		if result == types.MatchAgain {
+			again = true
+		}
+	}
+	if again {
+		return stream.EAGAIN
```
Comment (champly) [inline pkg/stream/xprotocol/factory.go:68]: If the protocol is unknown, always return `stream.EAGAIN` will fall into dead loop.

File: pkg/protocol/xprotocol/factory.go [hunk 2]
```diff
@@ -65,6 +65,20 @@ func GetMatcher(name types.ProtocolName) types.ProtocolMatch {
 	return matcherMap[name]
 }
 
+//GetMatchers return all matchers that was registered
+func GetMatchers() []types.ProtocolMatch {
```
Comment (champly) [inline pkg/protocol/xprotocol/factory.go:69]: Should build `matchers` with `RegisterMatcher`, each request build `matchers` will cost performance.

Comment (cadeeper) [reply pkg/protocol/xprotocol/factory.go:69]: Matchers may be changed by the caller, for this reason, I build it every request. I'll optimize it.

Comment (cadeeper) [reply pkg/protocol/xprotocol/factory.go:69]: > Should build `matchers` with `RegisterMatcher`, each request build `matchers` will cost performance.

Matchers may be changed by the caller, for this reason, I build it every request. I'll optimize it.

Comment (champly) [reply pkg/protocol/xprotocol/factory.go:69]: Why not return `matcherMap` direct?

Comment (cadeeper) [reply pkg/protocol/xprotocol/factory.go:69]: > Why not return `matcherMap` direct?

To protect matcherMap from change by the caller

File: pkg/stream/xprotocol/factory.go (hunk 1 above)
Comment (cadeeper) [reply pkg/stream/xprotocol/factory.go:68]: > If the protocol is unknown, always return `stream.EAGAIN` will fall into dead loop.

I don't understand. stream.EAGAIN only happened that subProtocolMatchers returns types.MatchAgain. if the protocol is unknown means each subProtocolMatchers will finally return types.MatchFailed. Then ProtocolMatch will return stream.FAILED

File: pkg/protocol/xprotocol/factory.go (hunk 2 above)
Comment (cadeeper) [reply pkg/protocol/xprotocol/factory.go:69]: > Why not return `matcherMap` direct?

If it's not necessary.I'll return matcherMap direct.

File: pkg/stream/xprotocol/factory.go [hunk 3]
```diff
@@ -49,5 +50,23 @@ func (f *streamConnFactory) CreateBiDirectStream(context context.Context, connec
 }
 
 func (f *streamConnFactory) ProtocolMatch(context context.Context, prot string, magic []byte) error {
+	subProtocolMatchers := xprotocol.GetMatchers()
```
Comment (taoyuanyuan) [inline pkg/stream/xprotocol/factory.go:53]: don't need configure `sub_protocol` by set "types.ContextSubProtocol"

Comment (cadeeper) [reply pkg/stream/xprotocol/factory.go:54]: done

Comment (cadeeper) [reply pkg/stream/xprotocol/factory.go:54]: > don't need configure `sub_protocol` by set "types.ContextSubProtocol"

done

**Comment by taoyuanyuan** [comment]
change base branch to master

**Review by taoyuanyuan — DISMISSED** [review]
LGTM

**Review by champly — APPROVED** [review]
LGTM

**Comment by taoyuanyuan** [comment]
it's wrong when configure upstreamProtocol auto.
https://github.com/mosn/mosn/blob/916254bc42db353adf9ea792347180de1d0d0f3b/pkg/stream/xprotocol/connpool.go#L100
https://github.com/mosn/mosn/blob/916254bc42db353adf9ea792347180de1d0d0f3b/pkg/stream/xprotocol/connpool_binding.go#L231

**Comment by taoyuanyuan** [comment]
we could delete `func getProtocol()`, and set `protocol.Xprotocol` directly

**Comment by cadeeper** [comment]
> we could delete `func getProtocol()`, and set `protocol.Xprotocol` directly

What will happen if it is not set to protocol.Xprotocol directly? this will set to the same as downstream's protocol if upstream protocol is AUTO:
https://github.com/mosn/mosn/blob/916254bc42db353adf9ea792347180de1d0d0f3b/pkg/proxy/downstream.go#L716

**Comment by taoyuanyuan** [comment]
> > we could delete `func getProtocol()`, and set `protocol.Xprotocol` directly
> 
> What will happen if we don't set to protocol.Xprotocol directly? this will set to the same as downstream's protocol if upstream protocol is AUTO:
> https://github.com/mosn/mosn/blob/916254bc42db353adf9ea792347180de1d0d0f3b/pkg/proxy/downstream.go#L716

The result is the same， the func getProtocol return `protocol.Xprotocol` certainly.

**Comment by cadeeper** [comment]
> > > we could delete `func getProtocol()`, and set `protocol.Xprotocol` directly
> > 
> > 
> > What will happen if we don't set to protocol.Xprotocol directly? this will set to the same as downstream's protocol if upstream protocol is AUTO:
> > https://github.com/mosn/mosn/blob/916254bc42db353adf9ea792347180de1d0d0f3b/pkg/proxy/downstream.go#L716
> 
> The result is the same， the func getProtocol return `protocol.Xprotocol` certainly.

done

**Review by taoyuanyuan — APPROVED** [review]
LGTM

**Comment by taoyuanyuan** [comment]
thx

**Pull request merged**